### ISSUE TITLE
Should oauth_xxxxx_eval functions return str or bytes?

### ISSUE THREAD

**Issue by piyueh** [OfflineIMAP/offlineimap3#57]
#### General informations

- system/distribution (with version): Arch Linux, kernel 5.11.1-arch1-1
- offlineimap version (`offlineimap -V`): 
    - from `offlineimap -V`: offlineimap v7.3.0, imaplib2 v3.06, Python v3.9.2, OpenSSL 1.1.1j  16 Feb 2021
    - it's actually offlineimap3 commit 1e7ef9e with patches from #55 and changes according to jazzband/imaplib2#15)
- Python version: 3.9
- server name or domain: gmail
- CLI options: None

#### Configuration file offlineimaprc

```
[general]
metadata = $XDG_DATA_HOME/offlineimap
accounts = xxx@gmail.com 
pythonfile = $XDG_CONFIG_HOME/offlineimap/offlineimap_funcs.py

[mbnames]
enabled = yes
filename = $XDG_CONFIG_HOME/offlineimap/mailboxes
header = "mailboxes "
peritem = "+%(accountname)s/%(foldername)s"
sep = " "
footer = "\n"
incremental = no

[Account xxx@gmail.com]
localrepository = xxx@gmail.com-local
remoterepository = xxx@gmail.com-remote
maxage = 60
synclabels = yes

[Repository xxx@gmail.com-local]
type = GmailMaildir
localfolders = $HOME/Sync/MailDir/xxx@gmail.com
sep = /
sync_deletes = no
restoreatime = no

[Repository xxx@gmail.com-remote]
type = Gmail
remoteuser = xxx@gmail.com
ssl = yes
sslcacertfile = /etc/ssl/certs/ca-certificates.crt
ssl_version = tls1_2
auth_mechanisms = XOAUTH2
oauth2_client_id_eval = get_gmail_client_id("xxx@gmail.com", "private-client.json")
oauth2_client_secret_eval = get_gmail_client_secret("xxx@gmail.com", "private-client.json")
oauth2_access_token_eval = get_gmail_token("xxx@gmail.com", "private-client.json")
oauth2_refresh_token_eval = get_gmail_refresh_token("xxx@gmail.com", "private-client.json")
expunge = yes
subscribedonly = no
createfolders = False
sync_deletes = yes
readonly = True
```

#### Python file `offlineimap_funcs.py`

See https://gist.github.com/piyueh/a2d65e095ea675a2c715ad42b7b61d10

#### Issue description

In the documentation in `offlineimap.conf`, it says the return values of the functions used by `oauth2_client_id_eval`, `oauth2_client_secret_eval`, `oauth2_access_token_eval`, and `oauth2_refresh_token_eval` should be `bytes`. However, returning `bytes` caused errors like this:

```
...
...
File "<redacted>/offlineimap/repository/IMAP.py", line 465, in getoauth2_client_id
  client_id = client_id.strip("\n")
TypeError: a bytes-like object is required, not 'str'
```

It is caused by that `client_id` is of `bytes` type, while the `"\n"` character is of `str` type. One possible fix is to use `b"\n"` instead. However, once I fixed the `"\n"` to `b"\n"` at line [431](https://github.com/OfflineIMAP/offlineimap3/blob/1e7ef9e7e6952f5d29ef0f5c25fd062798de55f3/offlineimap/repository/IMAP.py#L431), [448](https://github.com/OfflineIMAP/offlineimap3/blob/1e7ef9e7e6952f5d29ef0f5c25fd062798de55f3/offlineimap/repository/IMAP.py#L448), [465](https://github.com/OfflineIMAP/offlineimap3/blob/1e7ef9e7e6952f5d29ef0f5c25fd062798de55f3/offlineimap/repository/IMAP.py#L465), and [482](https://github.com/OfflineIMAP/offlineimap3/blob/1e7ef9e7e6952f5d29ef0f5c25fd062798de55f3/offlineimap/repository/IMAP.py#L482), a new error occurred:

```
*** Processing account xxx@gmail.com
Establishing connection to imap.gmail.com:993 (xxx@gmail.com-remote)
XOAUTH2 authentication failed: b'[AUTHENTICATIONFAILED] Invalid credentials (Failure)'
ERROR: All authentication types failed:
        XOAUTH2: b'[AUTHENTICATIONFAILED] Invalid credentials (Failure)'
*** Finished account 'xxx@gmail.com' in 0:00
ERROR: Exceptions occurred during the run!
ERROR: All authentication types failed:
        XOAUTH2: b'[AUTHENTICATIONFAILED] Invalid credentials (Failure)'

Traceback:
  File "<redacted>/offlineimap/accounts.py", line 298, in syncrunner
    self.__sync()
  File "<redacted>/offlineimap/accounts.py", line 374, in __sync
    remoterepos.getfolders()
  File "<redacted>/offlineimap/repository/IMAP.py", line 672, in getfolders
    imapobj = self.imapserver.acquireconnection()
  File "<redacted>/offlineimap/imapserver.py", line 588, in acquireconnection
    self.__authn_helper(imapobj)
  File "<redacted>/offlineimap/imapserver.py", line 453, in __authn_helper
    raise OfflineImapError("All authentication types "
```

Then, I reverted back to using `"\n"` and return `str` instead of `bytes` from those evaluation functions. Now it's working!!!

So now I'm confused. It seems these evaluation functions (for `oauth2_client_id_eval`, `oauth2_client_secret_eval`, `oauth2_access_token_eval`, and `oauth2_refresh_token_eval`) should return `str`, not `bytes`. But the documentation in `offlineimap.conf` says they should return `bytes`. Can anyone help or shed some light? Thanks!!

**Comment by sudipm-mukherjee** [comment]
@piyueh can you please paste your original `offlineimap_funcs.py` file also.

**Comment by piyueh** [comment]
@sudipm-mukherjee Hi, I pasted it to the Gist: https://gist.github.com/piyueh/a2d65e095ea675a2c715ad42b7b61d10

**Comment by piyueh** [comment]
I believe this is caused by the transition from Python 2 to Python 3. Handling of `str`, Unicode, and `bytes` is different in Python 2 and 3. Seems now the user-defined functions that return tokens do not have to return `bytes`.

**Comment by thekix** [comment]
Hello @piyueh 

Thanks a lot for your issue. Yes, the problem is related to Python2/Python3 change. In Python2 bytes and strings were more or less the same, but in Python3 are different.

Some questions/things:

1. If you return strings (str) in offlineimap_functions.py the problem is solved?
2. In the offlineimap_functions.py file the header is "#! /usr/bin/env python" and probably should be "#! /usr/bin/env python3"

The offlineimap_functions.py file seems to be fine and I am not sure if the best for offlineimap3 is using strings instead of bytes. Now, in offlineimap_functions.py, you are creating an object, with strings, and you are converting the strings to bytes for offlineimap3.

Probably is better avoid the conversion and return the string. IMHO, "Old" Python2 scripts cannot be supported in offlineimap3, and should be updated too (like offlineimap). If you need help to update this file, I can help. In this case, IMO, we should change the documentation.

What do you think?

Best Regards,
kix

PS. Thanks a lot for your help @sudipm-mukherjee

**Comment by piyueh** [comment]
Hi @thekix 

1. Yes, if I return `str`, the problem is solved.
2. I'm using Arch, and the default Python version in Arch is Python 3. So I don't explicitly use `python3` in the header.

I also believe the best way is to return `str` because that's how Python 3 works by default. Then, I think only the documentation should be changed.

**Comment by thekix** [comment]
Thanks a lot @piyueh 

I am trying to find it in the documentation, but I cannot find it. Where is it?

Thanks a lot,
Regards.
kix

**Comment by piyueh** [comment]
Hi @thekix, I think the fix is addressed in PR #77

**Comment by nicolas33** [comment]
Correct. Do never use the "go back in history" button in the browser while making PR on github...
